src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/LegacyCSProjPackageReferenceProject.cs

```diff
@@ -0,0 +1,341 @@
+            InternalMetadata.Add(NuGetProjectMetadataKeys.UniqueName, _project.UniqueName);
+            InternalMetadata.Add(NuGetProjectMetadataKeys.FullPath, _project.ProjectFullPath);
+        }
+
+        #region IDependencyGraphProject
+
+        public override string AssetsFilePath { get; }
+
+        public override string ProjectName { get; }
+
+        /// <summary>
+        /// Making this timestamp as the current time means that a restore with this project in the graph
+        /// will never no-op. We do this to keep this work-around implementation simple.
+        /// </summary>
+        public override DateTimeOffset LastModified => DateTimeOffset.Now;
+
+        public override string MSBuildProjectPath => _project.ProjectFullPath;
+
+        private IScriptExecutor ScriptExecutor
+        {
+            get
+            {
+                if (_scriptExecutor == null)
+                {
+                    _scriptExecutor = ServiceLocator.GetInstanceSafe<IScriptExecutor>();
+                }
+
+                return _scriptExecutor;
+            }
+        }
+
+        public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(DependencyGraphCacheContext context)
+        {
+            PackageSpec packageSpec;
+            if (context.PackageSpecCache.TryGetValue(_project.ProjectFullPath, out packageSpec))
+            {
+                return new[] { packageSpec };
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            packageSpec = await GetPackageSpec();
+            context.PackageSpecCache.Add(_project.ProjectFullPath, packageSpec);
+            return new[] { packageSpec };
+        }
+
+        public override async Task<bool> ExecuteInitScriptAsync(
+            PackageIdentity identity,
+            string packageInstallPath,
+            INuGetProjectContext projectContext,
+            bool throwOnFailure)
+        {
+            return
+                await
+                    ScriptExecutorUtil.ExecuteScriptAsync(identity, packageInstallPath, projectContext, ScriptExecutor,
+                        _project.DTEProject, throwOnFailure);
+        }
+
+        #endregion
+
+        #region NuGetProject
+
+        public override async Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            return GetPackageReferences(await GetPackageSpec());
+        }
+
+        public override async Task<Boolean> InstallPackageAsync(PackageIdentity packageIdentity, DownloadResourceResult downloadResourceResult, INuGetProjectContext nuGetProjectContext, CancellationToken token)
+        {
+            try
+            {
+                // We don't adjust package reference metadata from UI
+                await _project.AddOrUpdateLegacyCSProjPackageAsync(
```

Comment:
There is very little business logic in the adapter--this is specifically because the adapter exists *not to test the adapter but to test the project flavor". We don't need an adpater for an adapter's sake--DTE is fine. It exists to implement IEnvDTEProjectAdapter which can be mocked away in testing. I've pushed as much business logic as I can (including casting to more complex types) to the project layer in order to make it testable.